---
title: "scRNAseq_pipeline"
output: pdf_document
params:
  output: "default_output" 
  config_path: "default_config_path" 
---

# ENVIRONMENT SETUP

```{r env, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(reticulate)
use_condaenv("/w5home/bmoore/miniconda3/envs/scRNAseq_new", required = TRUE)#"/root/miniconda/envs/scrnaseq"
library(reticulate)
library(future)
library(future.apply)
library(furrr)
library(purrr)
library(dplyr)
library(Seurat)
library(patchwork)
library(scran)
library(BiocParallel)
library(DropletUtils)
library(cowplot)
library(harmony)
library(SoupX)
library(scDblFinder)
library(jsonlite)
library(rmarkdown)
library(tidyverse)
library(tibble)
library(stringr)
library(clustifyr)
library(scry)
library(ggplot2)
library(ComplexHeatmap)
#library(leidenAlg)
setwd("/w5home/bmoore/scRNAseq_library/sc_pipeline/src/")
output <- "/w5home/bmoore/scRNAseq_library/output/output_20250122_204811/"
# Load configuration and set output directory
config <- jsonlite::fromJSON(paste0(output, "config.json"), simplifyDataFrame = FALSE)
output_base_dir <- output
source("sc_pipeline_functions.R")
source("process_lane.R")
```
```{r}
# Assuming process_lane, combine_feature_plots, and split_sce_list_by_sample are defined in 'sc_pipeline_functions.R'
processed_data_list <- lapply(config$lanes, process_lane)
seurat_list <- purrr::map(processed_data_list, "seu")
sce_list <- purrr::map(processed_data_list, "sce")

feature_set <- list(feature1 = "nCount_RNA", feature2 = "nFeature_RNA")
combine_feature_plots(seurat_list, output_base_dir, feature_set, file_name_base = "post_soupx_qc")

sample_specific_sce_list <- split_sce_list_by_sample(sce_list)
clean_environment(list_to_remove = c("processed_data_list", "seurat_list", "feature_set"))

processed_samples_list <- list()
for (sample_name in names(sample_specific_sce_list)) {
  sample_data <- sample_specific_sce_list[[sample_name]]
  seurat_obj <- process_sample(sample_name, sample_data, output_base_dir, config)
  processed_samples_list[[sample_name]] <- seurat_obj
}

# Save the processed Seurat objects
saveRDS(processed_samples_list, file = paste0(output_base_dir, "processed_seurat_objects.rds"))
```
# if process didn't complete- read in object here
```{r}
seurat.obj <- readRDS(file = paste0(output, "seurat_obj_labeled.rds"))
print(seurat.obj)
```
# retry clustifyr
```{r}
clustifyR_obj <- annotate_with_clustifyR(seurat.obj, output)
print(clustifyR_obj)
```
# make list
```{r}
processed_samples_list <- list(labeled_seurat_obj = seurat.obj, clustifyR_obj = clustifyR_obj)
print(processed_samples_list)
```
```{r}
processed_samples_list2 <- list()
processed_samples_list2[["sample1"]] <- processed_samples_list
print(processed_samples_list2)
```
```{r}
seurat.objs <- readRDS(file = "/w5home/bmoore/scRNAseq_library/output/output_20240517_165517/processed_seurat_objects.rds")
print(seurat.objs)
```
# check running dimplot with phase
```{r}
pdf(paste0(output, "pca_after_cc_regression.pdf"), width = 8, height = 6)
    print(DimPlot(seurat.obj, group.by= "Phase"))
    dev.off()
```
# try rerunning score markers
```{r}
annot_df <- score_and_plot_markers(seurat.obj, output)
```
# annotate
```{r}
new_df_ordered <- annot_df[order(as.numeric(annot_df$Cluster)), ]
# Get new cluster names ordered by cluster number
new_cluster_ids <- new_df_ordered$Cell.type
names(new_cluster_ids) <- levels(seurat.obj)
seurat.obj$CellType <- Idents(seurat.obj)
head(seurat.obj@meta.data)
```
# switch back to seurat cluster as Idents
```{r}
Idents(seurat.obj) <- "seurat_clusters"
head(Idents(seurat.obj))
```
```{r}
seurat.obj <- annotate_clusters_and_save(seurat.obj, new_cluster_ids, output)
```
```{r}
# read in processed samples list
processed_samples_list<- readRDS(file = paste0(output_base_dir, "processed_seurat_objects.rds"))
print(processed_samples_list)
print(names(processed_samples_list))
```

```{r}
#if processed list doesn't read in , try batch-corrected objects
query.obj <- readRDS(file = paste0(output_base_dir, "GAMM_120/GAMM_120_batchcorr_seurat_obj.rds"))
ref.obj <- readRDS(file = paste0(output_base_dir, "Reh_205/Reh_205_batchcorr_seurat_obj.rds"))
```
```{r}
processed_samples_list <- list()
processed_samples_list[["Reh_205"]] <- ref.obj[[1]]
processed_samples_list[["GAMM_120"]] <- query.obj[[1]]
print(processed_samples_list)
```
```{r}
# Do cross species
source("sc_pipeline_functions.R")
if (!species_are_all_same(config)) {
  objs_list <- perform_orthologous_gene_analysis(processed_samples_list, config, output_base_dir)
}
```
```{r}
# read in subsetted data
query.seurat.subset <- readRDS(paste0(output_base_dir,"query_ortho-subset_seurat.rds"))
ref.seurat.subset <- readRDS(file = paste0(output_base_dir,"ref_ortho-subset_seurat.rds"))
print(query.seurat.subset)
print(ref.seurat.subset)
```
```{r}
query_cells <- colnames(query.seurat.subset)


# Add "_1" suffix to match harmony embeddings cell names
query_cells_modified <- paste0(query_cells, "_1")


# Subset the harmony embeddings to match the cells in the Seurat objects
harmony_embeddings_Q <- harmony_embeddings_Q[query_cells_modified, , drop = FALSE]


# Remove the "_1" suffix from rownames to match Seurat object cell names
rownames(harmony_embeddings_Q) <- query_cells


      # Create dimension reduction objects with matched embeddings
      query.seurat.subset[["harmony"]] <- CreateDimReducObject(
        embeddings = harmony_embeddings_Q,
        key = "harmony_",
        assay = DefaultAssay(query.seurat.subset)
      )

```

# starting with process_orthologous_objects
```{r}
# read in ortho.obj.list
objs_list <- readRDS(file = file.path(output_base_dir, "ortholog_objs_list.rds"))
objs_list

```
```{r}
  if (!is.null(objs_list)) {
    sample_names <- names(objs_list)
    if (length(sample_names) >= 2) {
      ref_name <- sample_names[1]
      query_name <- sample_names[2]
      
      message(paste("Processing reference:", ref_name))
      ref_list <- process_orthologous_objects(objs_list[[ref_name]], output_base_dir, config, ref_name)
      
      message(paste("Processing query:", query_name))
      query_list <- process_orthologous_objects(objs_list[[query_name]], output_base_dir, config, query_name)
    } else {
      warning("Insufficient samples for cross-species analysis")
    }
  } else {
    warning("No objects returned from orthologous gene analysis")
  }
```