---
title: "scRNAseq_pipeline"
output: pdf_document
params:
  output: "default_output" 
  config_path: "default_config_path" 
---

# ENVIRONMENT SETUP

```{r env, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(reticulate)
use_condaenv("/root/miniconda/envs/scrnaseq")
library(dplyr)
library(Seurat)
library(patchwork)
library(scran)
library(BiocParallel)
library(DropletUtils)
library(cowplot)
library(harmony)
library(SoupX)
library(scDblFinder)
library(purrr)
library(jsonlite)
library(rmarkdown)
library(tidyverse)
library(tibble)
library(stringr)
library(clustifyr)
library(ggplot2)
library(ComplexHeatmap)
output <- params$output
config <- jsonlite::fromJSON(params$config_path, simplifyDataFrame = FALSE)
s_output <- paste0("../", output, "/")
source("sc_pipeline_functions.R")
```

# Load data
```{r Load, SoupX and Obj Creation}
all_processed_data <- lapply(config$lanes, process_lane)
seurat_objs_list <- purrr::map(all_processed_data, "seu")
filtered_sce_list <- purrr::map(all_processed_data, "sce")
feature_set1 <- list(feature1 = "nCount_RNA", feature2 = "nFeature_RNA")
combine_feature_plots(seurat_objs_list, feature_set1, file_name_base = "post_soupx_qc")
sample_specific_list <- split_sce_list_by_sample(filtered_sce_list)
rm(all_processed_data, seurat_objs_list, feature_set1)
gc(full = TRUE)
```

# Sample Specific QC
```{r Sample Specific QC}
processed_seurat_objs <- list()
for (sample_name in names(sample_specific_list)) {
  # make output directory for each sample
  dir.create(paste0(sub_output, sample_name), showWarnings = FALSE)
  output <- paste0(output, sample_name, "/") # Update output path for each sample
  # 1. Doublet finder and lane merging
  lane_and_merged_seurat_obj <- run_scDblFinder_and_merge(sample_specific_list[[sample_name]])
  # 2. Mitochondrial gene filterin
  seurat_obj_mt_filtered <- filter_cells(lane_and_merged_seurat_obj, sample_name)
  rm(lane_and_merged_seurat_obj)
  gc(full = TRUE)
  normalized_seurat_obj <- normalize_data(seurat_obj_mt_filtered)
  rm(seurat_obj_mt_filtered)
  gc(full = TRUE)
  feature_selected_seurat_obj <- feature_selection(normalized_seurat_obj)
  rm(normalized_seurat_obj)
  gc(full = TRUE)
  scaled_seurat_obj <- scale_data(feature_selected_seurat_obj)
  rm(feature_selected_seurat_obj)
  gc(full = TRUE)
  dim_reduced_seurat_obj <- run_and_visualize_pca(scaled_seurat_obj)
  rm(scaled_seurat_obj)
  gc(full = TRUE)
  if (length(unique(dim_reduced_seurat_obj$orig.ident)) > 1) {
    batch_corrected_obj <- perform_batch_correction(dim_reduced_seurat_obj)
  } else {
    message("Skipping batch correction as orig.ident has only one level.")
    batch_corrected_obj <- dim_reduced_seurat_obj
  }
  rm(dim_reduced_seurat_obj)
  gc(full = TRUE)
  # add the batch corrected object to the list
  if (species_are_all_same(config)) {
    message("Species are consistent across all samples.")
    if (is.list(batch_corrected_obj)) {
      umap_seurat_obj <- run_umap(batch_corrected_obj$seurat_obj)
    } else {
      umap_seurat_obj <- run_umap(batch_corrected_obj)
    }
    # 9. Clustering
    clustered_seurat_obj <- perform_clustering(umap_seurat_obj)
    # 10. Find markers with Differential expressed features (genes)
    # Add your differential expression analysis code here
    # Store the final clustered object for this sample
    if (config$DE_method == "Seurat") {
      de_results <- find_differentially_expressed_features(clustered_seurat_obj)
      analyze_known_markers(clustered_seurat_obj, de_results, output)
    } else if (config$DE_method == "Scran") {
      annot_df <- score_and_plot_markers(clustered_seurat_obj, output)
    }
    if (config$score_and_plot_markers$known_markers) {
      new_df.ordered <- annot_df[order(as.numeric(annot_df$Cluster)), ]
      # get new cluster names ordered by cluster number
      new.cluster.ids <- new_df.ordered$Cell.type
      labeled_seurat_obj <- annotate_clusters_and_save(clustered_seurat_obj, new.cluster.ids, output)

      clustifyR_obj <- annotate_with_clustifyR(clustered_seurat_obj, output)
      # append both annotated objects to the list
      processed_seurat_objs[[sample_name]] <- list(labeled_seurat_obj = labeled_seurat_obj, clustifyR_obj = clustifyR_obj)
    }
  } else {
    message("Species are not consistent across all samples. Initiating orthologous gene analysis.")
    # save the batch corrected object to the list for orthologous gene analysis
    processed_seurat_objs[[sample_name]] <- batch_corrected_obj
  }
}
```

# save the processed seurat objects
```{r Save the processed seurat objects}
saveRDS(processed_seurat_objs, file = paste0(output, "processed_seurat_objs.rds"))

```
# Orthologous Gene Analysis
```{r Orthologous Gene Analysis}
if (!species_are_all_same(config)) {

    # determine the species of the first sample in the list
    categorized_samples <- categorize_samples_by_species(processed_seurat_objs, config)
    ref_obj <- categorized_samples$ref_objects[[1]]
    query_obj <- categorized_samples$query_objects[[1]]
    ref_obj <- run_scDblFinder_and_merge(ref_obj)
    query_obj <- run_scDblFinder_and_merge(query_obj)
    ref_project <- ref_obj$orig.ident[[1]]
    query_project <- query_obj$orig.ident[[1]]
    ref_project <- sub("_lane.*", "", ref_project)
    query_project <- sub("_lane.*", "", query_project)
    project_names <- c(ref_project, query_project)
    ref_obj <- filter_cells(ref_obj, ref_project)
    query_obj <- filter_cells(query_obj, query_project)
    objs.list <- ortholog_subset(ref_obj, query_obj, project_names)
    saveRDS(objs.list, file = paste0(output, "ortholog_objs_list.rds"))
  }
}
```


