---
title: "scRNAseq_pipeline"
output: pdf_document
params:
  output: "default_output" 
  config_path: "default_config_path" 
---

# ENVIRONMENT SETUP

```{r env, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(reticulate)
use_condaenv("/root/miniconda/envs/scrnaseq")
library(dplyr)
library(Seurat)
library(patchwork)
library(scran)
library(BiocParallel)
library(DropletUtils)
library(cowplot)
library(harmony)
library(SoupX)
library(scDblFinder)
library(purrr)
library(jsonlite)
library(rmarkdown)
library(tidyverse)
use_condaenv("/root/miniconda/envs/scrnaseq")
library(tibble)
library(stringr)
library(clustifyr)
library(ggplot2)
library(ComplexHeatmap)
output <- params$output
config <- jsonlite::fromJSON(params$config_path, simplifyDataFrame = FALSE)
output <- paste0("../", output, "/")
source("sc_pipeline_functions.R")
```

# Load data
```{r Load, SoupX and Obj Creation}
# Process the data
all_processed_data <- lapply(config$lanes, process_lane)
seurat_objs_list <- purrr::map(all_processed_data, "seu")
filtered_sce_list <- purrr::map(all_processed_data, "sce")
feature_set1 <- list(feature1 = "nCount_RNA", feature2 = "nFeature_RNA")
combine_feature_plots(seurat_objs_list, feature_set1, file_name_base = "post_soupx_qc")
rm(all_processed_data, seurat_objs_list, feature_set1)
gc(full = TRUE)
```

# Dbl Remover
```{r scDbl Finder and lane merging}
# Run doublet finder and merge lanes
lane_and_merged_seurat_obj <- run_scDblFinder_and_merge(filtered_sce_list)
rm(filtered_sce_list)
gc(full = TRUE)
```

# Mitochondrial gene filtering
```{r MT Gene Filtering}
seurat_obj_mt_filtered <- filter_cells(lane_and_merged_seurat_obj)
rm(lane_and_merged_seurat_obj)
gc(full = TRUE)
```

# Normalization
```{r Normalization}
normalized_seurat_obj <- normalize_data(seurat_obj_mt_filtered)
rm(seurat_obj_mt_filtered)
gc(full = TRUE)
```

# Feature selection
```{r Feature Selection}
feature_selected_seurat_obj <- feature_selection(normalized_seurat_obj)
rm(normalized_seurat_obj)
gc(full = TRUE)
```

# Scaling data
```{r Scaling Data}
scaled_seurat_obj <- scale_data(feature_selected_seurat_obj)
rm(feature_selected_seurat_obj)
gc(full = TRUE)
```

# Dimensional reduction
```{r Dimensional Reduction}
dim_reduced_seurat_obj <- run_and_visualize_pca(scaled_seurat_obj)
rm(scaled_seurat_obj)
gc(full = TRUE)
```

# Batch correction
```{r Batch Correction}
# Check the number of unique levels in orig.ident
if (length(unique(dim_reduced_seurat_obj$orig.ident)) > 1) {
  # Perform batch correction if orig.ident has more than one level
  batch_corrected_obj <- perform_batch_correction(dim_reduced_seurat_obj)
} else {
  # Skip batch correction and use the original object
  batch_corrected_obj <- dim_reduced_seurat_obj
  message("Skipping batch correction as orig.ident has only one level.")
}

# Remove the original object to free up memory
rm(dim_reduced_seurat_obj)
gc(full = TRUE)
```

# UMAP
```{r UMAP}
# Check if batch_corrected_obj is a list (i.e., batch correction was performed)
if (is.list(batch_corrected_obj)) {
  # Run UMAP on the Seurat object returned by perform_batch_correction
  umap_seurat_obj <- run_umap(batch_corrected_obj$seurat_obj)
} else {
  # Run UMAP on the original Seurat object (i.e., batch correction was skipped)
  umap_seurat_obj <- run_umap(batch_corrected_obj)
}

# Remove the batch_corrected_obj to free up memory
rm(batch_corrected_obj)
gc(full = TRUE)
```

# Clustering 
```{r Clustering}
clustered_seurat_obj <- perform_clustering(umap_seurat_obj)
rm(umap_seurat_obj)
gc(full = TRUE)
```

# Find markers with Differential expressed features (genes)
```{r Differential Expression}
if (config$DE_method == "Seurat") {
  de_results <- find_differentially_expressed_features(clustered_seurat_obj)
  analyze_known_markers(clustered_seurat_obj, de_results, output)
} else if (config$DE_method == "Scran") {
  annot_df <- score_and_plot_markers(clustered_seurat_obj, output)
}
```
# save the clustered object
```{r Save Final Object}
saveRDS(clustered_seurat_obj, file = paste0(output, "clustered_seurat_obj.rds"))
```
# Both annotation methods
```{r manual_annot}
# order clusters from annotation data frame
new_df.ordered <- annot_df[order(as.numeric(annot_df$Cluster)), ]
# get new cluster names ordered by cluster number
new.cluster.ids <- new_df.ordered$Cell.type
# annotate
labeled_seurat_obj <- annotate_clusters_and_save(clustered_seurat_obj, new.cluster.ids, output)

clustifyR_obj <- annotate_with_clustifyR(clustered_seurat_obj, output)
```
