# load packages
```{r load_packages}
knitr::opts_chunk$set(echo = TRUE)
# conda activate scRNAseq_best
#devtools::install_github("powellgenomicslab/scPred")
library(scPred)
library(Seurat)
library(magrittr)
library(harmony)
library(rmarkdown)
library(jsonlite)
BETHS_GAMM_DATA_DIR <- "/w5home/bmoore/scRNAseq/GAMM/"
use_condaenv("scRNAseq_best")
setwd(paste0(BETHS_GAMM_DATA_DIR, "human_data/reh_cellrep_2020/"))
timestamp <- format(Sys.time(), "%Y%m%d_%H%M%S")
output <- paste0("output_scPred_", timestamp)
dir.create(output, showWarnings = FALSE)
file.copy("/w5home/bmoore/scRNAseq_library/config.json", file.path(output, "config.json"))
config <- fromJSON(file.path(output, "config.json"))
output <- paste0(output, "/")
source("/w5home/bmoore/scRNAseq_library/src/sc_pipeline_functions.R")
```
# run preprocess_crossspecies.Rmd first to get the data preprocessed the same way
# get reference and query data
```{r get_data}
human_D205.seurat <- readRDS(file = "output_20230711_155556/human_D205_umap.rds")
gamm <- readRDS(file = "output_20230711_155556/GAMM_S2_norm.rds")
```
# show umap of ref data
```{r show_umap_ref}
# we don't do clustering because clusters already annotated
DimPlot(human_D205.seurat, group.by = "type", label = TRUE, repel = TRUE)

 # Save UMAP clusters plot
  pdf(paste0(output, "umap_ref.pdf"), width = 8, height = 6)
  umap_clusters <- DimPlot(human_D205.seurat, group.by = "type", 
  reduction = "umap", label = TRUE, pt.size = .1)
  print(umap_clusters)
  dev.off()
# check metadata
table(human_D205.seurat$type)
# subset if needed
human_D205.seurat<- subset(human_D205.seurat, subset = type != c("AC2","T2"))
human_D205.seurat<- subset(human_D205.seurat, subset = type != c('AC2','T2'))
human_D205.seurat<- subset(human_D205.seurat, subset = type != c('AC2'))
table(human_D205.seurat$type)
```

# training the classifyer
```{r train_classifyer}
# getFeatureSpace will create a scPred object stored in the @misc slot. This 
# object will contained all required information to classify cells.
human_D205.seurat <- getFeatureSpace(human_D205.seurat, "type")

# train models for each cell type
# cv has to be set BELOW min class
human_D205.seurat <- trainModel(human_D205.seurat, resampleMethod = "cv",
                                number = 10, seed = 704)
# if you get this error:
# There were missing values in resampled performance measures.
# try: 1. make sure no NAs in data 2. decreasing the cvs, 3. down or up sample to make classes even
```
```{r check_classifyer}
# get training prob for each cell-cell type
get_probabilities(human_D205.seurat) %>% head()
# use get_scpred method to retrieve the scPred object from the Seurat object.
# this gives stats on each cell type prediction
get_scpred(human_D205.seurat)
model_stats <- get_scpred(human_D205.seurat)
# visualize cell type probabilties
pdf(paste0(output, "celltype_model_prob_svmpoly.pdf"), width = 8, height = 6)
plot_probabilities(human_D205.seurat)
dev.off()
```
# OPTIONAL: can try other models from caret:
# https://topepo.github.io/caret/available-models.html
# pass in the model parameter
# can reclassify a subset of cells (like those that didn't work well)
```{r train_classifyer2}
# other models tried: rf, logreg, xgbTree, mda, AdaBoost.M1
human_D205.seurat <- trainModel(human_D205.seurat, model = 'AdaBoost.M1', 
                                   resampleMethod = "cv",
                                   number = 10, seed = 704,
                                   reclassify = c("T1/T3","Prog/Glia","RGC"))
get_scpred(human_D205.seurat)
model_stats_ada <- get_scpred(human_D205.seurat)
pdf(paste0(output, "celltype_model_prob_adaboost.pdf"), width = 8, height = 6)
plot_probabilities(human_D205.seurat)
dev.off()
```
```{r train_classifyer3}
# other models 'glm'
human_D205.seurat <- trainModel(human_D205.seurat, model = "glm", 
                                   resampleMethod = "cv",
                                   number = 10, seed = 704,
                                   reclassify = "T1/T3")
get_scpred(human_D205.seurat)
pdf(paste0(output, "celltype_model_prob_glm.pdf"), width = 8, height = 6)
plot_probabilities(human_D205.seurat)
dev.off()

```
# get final model
```{r train_classifyer4}
# reclassify prog/glia and rgc with svm-radial
human_D205.seurat <- trainModel(human_D205.seurat, model = "svmRadial", 
                                   resampleMethod = "cv",
                                   number = 10, seed = 704,
                                   reclassify = c("Prog/Glia","RGC","T1/T3"))
get_scpred(human_D205.seurat)
pdf(paste0(output, "celltype_model_prob_svmRadial.pdf"), width = 8, height = 6)
plot_probabilities(human_D205.seurat)
dev.off()
```
# save reference seurat object
```{r save_ref}
saveRDS(human_D205.seurat, file = "human_D205_models2.rds")
```
# if you need to read objects back in
```{r read_in}
human_D205.seurat <- readRDS(file = "human_D205_models2.rds")
gamm <- readRDS(file = "output_20230711_155556/GAMM_S2_norm.rds")
```
# classify cells
```{r classify_cells}
# An important requirement for classifying cells is using the same normalization 
# method for both the reference and the query datasets.
# query data was normalized in preprocess_crossspecies.Rmd
# scPred now uses Harmony to align the query data onto the training low-dimensional 
# space used as reference. Once the data is aligned, cells are classified using the 
# pre-trained models.
# predict query
gamm <- scPredict(gamm, human_D205.seurat)

#plot the classifications over the aligned data.
pdf(paste0(output, "celltype_prediction_1.pdf"), width = 8, height = 6)
DimPlot(gamm, group.by = "scpred_prediction", reduction = "scpred")
dev.off()
# run UMAP using the aligned data as an input
gamm <- RunUMAP(gamm, reduction = "scpred", dims = 1:30)
# plot the predicted labels for each cell type over the UMAP:
pdf(paste0(output, "celltype_prediction_1-umap.pdf"), width = 8, height = 6)
DimPlot(gamm, group.by = "scpred_prediction", label = TRUE, repel = TRUE)
dev.off()
```
# visualize probabilities
```{r visualize_prob}
#  probabilities of each cell in the @meta.data slot of the query Seurat object.
# get labels
labels<- as.vector(unique(colnames(gamm@meta.data)))
labels<-labels[! labels %in% c('orig.ident', 'nCount_RNA', 'nFeature_RNA', 
        'scpred_max', 'scpred_prediction', 'scpred_no_rejection','sizeFactors',
        "scDblFinder.score","percent.mt","CellType","ident","scDblFinder.weighted",
        "RNA_snn_res.0.5","scDblFinder.class","scDblFinder.cxds_score","seurat_clusters")]
# visualize the probabilities over the UMAP plot:
pdf(paste0(output, "celltype_prediction_featureplot.pdf"), width = 11, height = 9)
print(FeaturePlot(gamm, labels, keep.scale = "all"))
dev.off()
```
# verify performance with manual annotation
```{r verify_performance}
# add in metadata from previous analysis
metadata.gamm <- read.csv("gamm_manual_annot_metadata_c0.14.txt", row.names = 1, 
            header = TRUE, sep = "\t")
# check metadata with query data
colnames(gamm@assays$RNA@data)[1:10]
rownames(metadata.gamm)[1:10]
# remove "_1" from metadata
rownames(metadata.gamm) <- sub("_1", "", rownames(metadata.gamm))
rownames(metadata.gamm)[1:10]
# add metadata to query data
gamm <- AddMetaData(gamm, metadata.gamm)
# verify model performance in query data
query.ref.data.table<- as.data.frame(crossTab(gamm, "CellType", "scpred_prediction"))
# check out proportion of cells
query.ref.data.table2<- as.data.frame(crossTab(gamm, "CellType", "scpred_prediction", output = "prop"))
write.table(query.ref.data.table, file = paste0(output, "query_ref_data_table_c0.14.txt"), sep = "\t")
write.table(query.ref.data.table2, file = paste0(output, "query_ref_data_table_proportion_c0.14.txt"), sep = "\t")
# get query umap with orignal labels
pdf(paste0(output, "celltype_originallabels-umap_c0.14.pdf"), width = 8, height = 6)
DimPlot(gamm, group.by = "CellType", repel = TRUE, label=TRUE)+ NoLegend()
dev.off()
```
# save query seurat object
```{r save_query}
saveRDS(gamm, file = "GAMM_S2_scpred_models_c0.14.rds")
```
```{r access_classifiers}
# access the classifiers
classifiers<- get_classifiers(human_D205.seurat)

saveRDS(classifiers, file = paste0(output, "classifiers.rds"))

# for (i in 1:length(classifiers)) {
#         print(classifiers[i])
#         write(names(classifiers[i]), file= paste0(output, "classifiers.txt"),
#         sep="\n", append=TRUE)
#         write(classifiers[[i]], file= paste0(output, "classifiers.txt"),
#         sep="\n", append=TRUE)

# }
# Each model can be normally treated using the caret enviroment.
# caret::plot.train(get_classifiers(human_D205.seurat)[["AC"]])
```