---
title: "preprocess_for_cross-species"
author: "Beth Moore"
date: "2023-07-07"
output: html_document
---
# Preprocessing for cross species analysis
# to have cross species, we need ortholog genes
# we then need to subset both reference and query by these genes
# reference data will also be subseted by its metadata

# Load libraries, set wd, and source functions
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(dplyr)
library(Seurat)
library(patchwork)
library(scran)
library(BiocParallel)
library(DropletUtils)
library(cowplot)
library(harmony)
library(SoupX)
library(scDblFinder)
library(reticulate)
library(purrr)
library(jsonlite)
library(rmarkdown)
BETHS_GAMM_DATA_DIR <- "/w5home/bmoore/scRNAseq/GAMM/"
use_condaenv("scRNAseq_best")
setwd(paste0(BETHS_GAMM_DATA_DIR, "human_data/cowan_cell_2020/"))
timestamp <- format(Sys.time(), "%Y%m%d_%H%M%S")
output <- paste0("output_preprocess", timestamp)
dir.create(output, showWarnings = FALSE)
file.copy("/w5home/bmoore/scRNAseq_library/config.json", file.path(output, "config.json"))
config <- fromJSON(file.path(output, "config.json"))
output <- paste0(output, "/")
source("/w5home/bmoore/scRNAseq_library/sc_pipeline/src/sc_pipeline_functions.R")

```
# Read in data
```{r data}
# refernce data
exp.matrix1 <- read_aligned_data(
    paste0(BETHS_GAMM_DATA_DIR,"human_data/cowan_cell_2020/output_F49i-N-B7/Solo.out/GeneFull/"),
    "F49i-N-B7",output
    )
exp.matrix2 <- read_aligned_data(
    paste0(BETHS_GAMM_DATA_DIR,"human_data/cowan_cell_2020/output_IMR90-4-DL-01/Solo.out/GeneFull/"),
    "IMR90-4-DL-01",output
    )

# load gamm pig data sample 2
gamm.data <- read_aligned_data(
    paste0(BETHS_GAMM_DATA_DIR,"output_S2_mm_mt/Solo.out/GeneFull/"),
    "gamm.data.S2",output
    )
```

# run soupx
```{r soupx}
all_aligned_data <- list(exp.matrix1,exp.matrix2,gamm.data)
soupX_objs <- purrr::map(all_aligned_data, ~prep_seurat_and_soupX(data.raw = .x$raw, data = .x$filtered, project = .x$project), .names = purrr::map_chr(all_data, 'project'))
# write out data
DropletUtils:::write10xCounts(paste0(output,"cowan_soupX_filt_F49i.mtx"), soupX_objs[[1]]$out)
DropletUtils:::write10xCounts(paste0(output,"cowan_soupX_filt_IMR90-4.mtx"), soupX_objs[[2]]$out)
DropletUtils:::write10xCounts(paste0(output,"gamm_soupX_filt_S2.mtx"), soupX_objs[[3]]$out)
ref1.matrix.soupx<- soupX_objs[[1]]$out
ref2.matrix.soupx<- soupX_objs[[2]]$out
gamm.matrix.soupx<- soupX_objs[[3]]$out
```
# Create Seurat objects and convert to sce objects for doublet removal
```{r}
project_names <- purrr::map_chr(all_aligned_data, 'project')
feature_set1 <- list(feature1 = 'nCount_RNA', feature2 = 'nFeature_RNA')
filtered_list_of_objs <- purrr::map2(soupX_objs, project_names, ~create_seurat_and_sce(out = .x$out, project = .y, feature_set = feature_set1))

```
# Doublet removal and lane merging
```{r}
combine_feature_plots(filtered_list_of_objs, feature_set1, file_name_base = "post_soupx_qc")
filtered_sce_list <- list(filtered_list_of_objs[[1]]$sce, filtered_list_of_objs[[2]]$sce, filtered_list_of_objs[[3]]$sce)
lane_and_merged_seurat_obj <- run_scDblFinder_and_merge(filtered_sce_list)
print(unique(lane_and_merged_seurat_obj$orig.ident))
```
# get sparse count matrices
```{r}
obj.list <- SplitObject(lane_and_merged_seurat_obj, split.by = "orig.ident")
human_ref_F49i<- obj.list$`F49i-N-B7`
human_ref_IMR90.4<- obj.list$`IMR90-4-DL-01`
pig_query<- obj.list$gamm.data.S2
dim(human_ref_F49i@assays$RNA@counts)
dim(human_ref_F49i@assays$RNA@data)
dim(human_ref_IMR90.4@assays$RNA@counts)
dim(human_ref_IMR90.4@assays$RNA@data)
dim(pig_query@assays$RNA@counts)
dim(pig_query@assays$RNA@data)
all.equal(pig_query@assays$RNA@counts,
pig_query@assays$RNA@data)
all.equal(human_ref_F49i@assays$RNA@counts,
human_ref_F49i@assays$RNA@data)
all.equal(human_ref_IMR90.4@assays$RNA@counts,
human_ref_IMR90.4@assays$RNA@data)
ref.matrix1.flt <- human_ref_F49i@assays$RNA@data
ref.matrix2.flt <- human_ref_IMR90.4@assays$RNA@data
gam.matrix.flt <- pig_query@assays$RNA@data
# write out data
DropletUtils:::write10xCounts(paste0(output,"cowan_dblfinder_filt_F49i.mtx"), ref.matrix1.flt)
DropletUtils:::write10xCounts(paste0(output,"cowan_dblfinder_filt_IMR90.4.mtx"), ref.matrix2.flt)
DropletUtils:::write10xCounts(paste0(output,"gamm_dblfinder_filt_S2.mtx"), gam.matrix.flt)
rm(ref1.matrix.soupx, ref2.matrix.soupx,gamm.matrix.soupx)
```
# read in orthologs
```{r orthologs}
# Reminder: these are assumed to be in the working directory...
orthologs <- read.csv2(paste0(BETHS_GAMM_DATA_DIR,"human_data/reh_cellrep_2020/biomart_export_human-pig_20230807_one-to-one.txt"), sep="\t", header=TRUE)
```
# subset matrices
```{r subset}
# subset genes so they match only orthologous ones
human.orthos <- as.vector(orthologs$human.gene.name)
ref.matrix1.flt<- ref.matrix1.flt[rownames(ref.matrix1.flt) %in% human.orthos, ]
ref.matrix2.flt<- ref.matrix2.flt[rownames(ref.matrix2.flt) %in% human.orthos, ]
# subset orthologs to they only match expr matrix
orthologs1 <- orthologs[
  orthologs$human.gene.name %in% rownames(ref.matrix1.flt) &
  orthologs$human.gene.name %in% rownames(ref.matrix2.flt), 
]
human.orthos <- as.vector(orthologs1$human.gene.name)
# check length to make sure they are the same
length(rownames(ref.matrix1.flt))
length(rownames(ref.matrix2.flt))
length(orthologs1$human.gene.name)
```
# switch to pig genes
```{r pig-genes}
# reorder based on exp matrix genes
ind_reorder <- match(rownames(ref.matrix1.flt),orthologs1$human.gene.name)
orthologs.reorder<- orthologs1[ind_reorder,]
# now replace rownames of exp matrix with pig genes
pig.orthos <- as.vector(orthologs.reorder$pig.gene.name)
head(row.names(ref.matrix1.flt))
row.names(ref.matrix1.flt) <- pig.orthos
# check exp matrix row names
head(row.names(ref.matrix1.flt))
length(rownames(ref.matrix1.flt))
# do for next reference
ind_reorder <- match(rownames(ref.matrix2.flt),orthologs1$human.gene.name)
orthologs2.reorder<- orthologs1[ind_reorder,]
pig.orthos2 <- as.vector(orthologs2.reorder$pig.gene.name)
head(row.names(ref.matrix2.flt))
row.names(ref.matrix2.flt) <- pig.orthos2
head(row.names(ref.matrix2.flt))
length(rownames(ref.matrix2.flt))
```
# load pig data and subset so both pig and human are the same
```{r pig-data}

# subset pig data so that there is same genes
dim(gam.matrix.flt)
gam.matrix.flt<- gam.matrix.flt[rownames(gam.matrix.flt) %in% pig.orthos, ]
dim(gam.matrix.flt)

# subset human refernce based on gamm data
ref.matrix1.flt<- ref.matrix1.flt[rownames(ref.matrix1.flt) %in% rownames(gam.matrix.flt), ]
dim(ref.matrix1.flt)
ref.matrix2.flt<- ref.matrix2.flt[rownames(ref.matrix2.flt) %in% rownames(gam.matrix.flt), ]
dim(ref.matrix2.flt)
DropletUtils:::write10xCounts(paste0(output,"ref1.filtered.exp"), ref.matrix1.flt)
DropletUtils:::write10xCounts(paste0(output,"ref2.filtered.exp"), ref.matrix2.flt)
DropletUtils:::write10xCounts(paste0(output,"gamm.data.filtered.exp"), gam.matrix.flt)
#ref.matrix.flt <- Read10X("ref.filtered.exp/")
#gam.matrix.flt <- Read10X("gamm.data.filtered.exp/")
```
# filter out duplicated genes in reference
```{r filter}
# get rid of duplicates by averaging them
# Aggregate rows based on row names
ref.matrix1.flt <- aggregate(ref.matrix1.flt, by = list(rownames(ref.matrix1.flt)), FUN = mean)
ref.matrix2.flt <- aggregate(ref.matrix2.flt, by = list(rownames(ref.matrix2.flt)), FUN = mean)
# use column 1 as rownames
rownames(ref.matrix1.flt)<- ref.matrix1.flt$Group.1
ref.matrix1.flt <- select(ref.matrix1.flt,  -c("Group.1"))
rownames(ref.matrix2.flt)<- ref.matrix2.flt$Group.1
ref.matrix2.flt <- select(ref.matrix2.flt,  -c("Group.1"))
# turn back to sparse matrix
#install.packages("Matrix", repos="http://R-Forge.R-project.org")
library(Matrix)
ref.matrix1.flt<- as.matrix(ref.matrix1.flt)
ref.matrix1.flt <- Matrix(ref.matrix1.flt, sparse = TRUE)
dim(ref.matrix1.flt)
ref.matrix2.flt<- as.matrix(ref.matrix2.flt)
ref.matrix2.flt <- Matrix(ref.matrix2.flt, sparse = TRUE)
dim(ref.matrix2.flt)
```
```{r}
DropletUtils:::write10xCounts(paste0(output,"ref1.filtered2.exp"), ref.matrix1.flt)
DropletUtils:::write10xCounts(paste0(output,"ref2.filtered2.exp"), ref.matrix2.flt)
```
# get metadata
```{r metadata}
# read in meta.data
# remember the set working directory in chunk 1, to imform this path
metadata <- read.csv2("meta_30weeks_mod.txt", sep="\t", header=TRUE, row.names=1)
# subset by D205 data
metadata1 <- subset(metadata, metadata$source=="f49b7")
metadata1<- as.data.frame(metadata1)
complete.cases(metadata1)
metadata1<-na.omit(metadata1)
metadata2 <- subset(metadata, metadata$source=="imr90")
metadata2<- as.data.frame(metadata2)
complete.cases(metadata2)
metadata2<-na.omit(metadata2)
# get numbers for each celltype
table(metadata1$cell_type2)
table(metadata2$cell_type2)
# replace underscore with dash
#row.names(metadata) <- sub("_1", "", row.names(metadata))
metadata.rows <-as.vector(row.names(metadata))
dim(metadata1)
dim(ref.matrix1.flt)
dim(metadata2)
dim(ref.matrix2.flt)
```
# create seurat object
```{r} 
human_F49B7.seurat<- CreateSeuratObject(ref.matrix1.flt, project = "F49B7", assay = "RNA",
                                       min.cells = 3, min.features = 200)
human_F49B7.seurat<-AddMetaData(object=human_F49B7.seurat, metadata=metadata1)
human_IMR90.seurat<- CreateSeuratObject(ref.matrix2.flt, project = "IMR90", assay = "RNA",
                                       min.cells = 3, min.features = 200)
human_IMR90.seurat<-AddMetaData(object=human_IMR90.seurat, metadata=metadata2)
gamm.S2.seurat<- CreateSeuratObject(gam.matrix.flt, project = "gamm_S2", assay = "RNA",
                                       min.cells = 3, min.features = 200)
```
# save seurat object
```{r}
saveRDS(human_F49B7.seurat, file = paste0(output,"human_F49B7_ortholog.rds"))
saveRDS(human_IMR90.seurat, file = paste0(output,"human_IMR90_ortholog.rds"))
saveRDS(gamm.S2.seurat, file = paste0(output,"GAMM_S2_ortholog.rds"))
```
# Mitochondrial gene filtering
```{r}
human_F49B7.seurat_mt_filtered <- filter_cells(human_F49B7.seurat)
human_IMR90.seurat_mt_filtered <- filter_cells(human_IMR90.seurat)
gamm.S2.seurat_mt_filtered <- filter_cells(gamm.S2.seurat)
```
# need to check if genes are the same after filtering, if not subset again
```{r}
human_F49B7.seurat.exp.mtx<- GetAssayData(human_F49B7.seurat_mt_filtered, slot="data")
human_IMR90.seurat.exp.mtx<- GetAssayData(human_IMR90.seurat_mt_filtered, slot="data")
gamm.S2.seurat.exp.mtx<- GetAssayData(gamm.S2.seurat_mt_filtered, slot="data")
dim(human_F49B7.seurat.exp.mtx)
dim(human_IMR90.seurat.exp.mtx)
dim(gamm.S2.seurat.exp.mtx)

 # subset gamm data based on human data
gamm.S2.seurat.exp.mtx<- gamm.S2.seurat.exp.mtx[rownames(gamm.S2.seurat.exp.mtx)
    %in% rownames(human_F49B7.seurat.exp.mtx), ]
gamm.S2.seurat.exp.mtx<- gamm.S2.seurat.exp.mtx[rownames(gamm.S2.seurat.exp.mtx)
    %in% rownames(human_IMR90.seurat.exp.mtx), ]
# subset human refernce based on gamm data
human_F49B7.seurat.exp.mtx<- human_F49B7.seurat.exp.mtx[rownames(human_F49B7.seurat.exp.mtx)
 %in% rownames(gamm.S2.seurat.exp.mtx), ]
human_IMR90.seurat.exp.mtx<- human_IMR90.seurat.exp.mtx[rownames(human_IMR90.seurat.exp.mtx)
 %in% rownames(gamm.S2.seurat.exp.mtx), ]

dim(human_F49B7.seurat.exp.mtx)
dim(human_IMR90.seurat.exp.mtx)
dim(gamm.S2.seurat.exp.mtx)
```
# set data in new seurat object
```{r}
human_F49B7.seurat_mt_filtered_subset<- CreateSeuratObject(human_F49B7.seurat.exp.mtx, 
    project = "human_F49B7", assay = "RNA")
human_F49B7.seurat_mt_filtered_subset<-AddMetaData(object=
    human_F49B7.seurat_mt_filtered_subset, metadata=metadata1)
human_IMR90.seurat_mt_filtered_subset<- CreateSeuratObject(human_IMR90.seurat.exp.mtx, 
    project = "human_IMR90", assay = "RNA")
human_IMR90.seurat_mt_filtered_subset<-AddMetaData(object=
    human_IMR90.seurat_mt_filtered_subset, metadata=metadata2)
gamm.S2.seurat_mt_filtered_subset<- CreateSeuratObject(gamm.S2.seurat.exp.mtx, 
    project = "gamm_S2", assay = "RNA")

dim(human_F49B7.seurat_mt_filtered_subset)
dim(human_IMR90.seurat_mt_filtered_subset)
dim(gamm.S2.seurat_mt_filtered_subset)
```

# Normalization
```{r}
human_F49B7.seurat_norm <- normalize_data(human_F49B7.seurat_mt_filtered_subset)
human_IMR90.seurat_norm <- normalize_data(human_IMR90.seurat_mt_filtered_subset)
gamm.S2.seurat_norm <- normalize_data(gamm.S2.seurat_mt_filtered_subset)
saveRDS(human_F49B7.seurat_norm, file = paste0(output,"human_F49B7_norm.rds"))
saveRDS(human_IMR90.seurat_norm, file = paste0(output,"human_IMR90_norm.rds"))
saveRDS(gamm.S2.seurat_norm, file = paste0(output,"GAMM_S2_norm.rds"))
```
## further steps: 
# for scPRED: reference undergoes feature selection, scaling, and dimensional reduction, query does not
#    # if doing a test set for scPred, stop here and load normalized data into scPred. 
#    # Scpred will split the reference into training and testing, then do the rest of the steps.
# for Seurat integration or mapping: query also undergoes feature selection

# Feature selection
```{r}
#gamm.S2.seurat_norm<-readRDS(file = "output_20230711_155556/GAMM_S2_norm.rds")
#human_D205.seurat_norm<-readRDS(file = "output_20230711_155556/human_D205_norm.rds")
gamm.S2.seurat_feat_select <- feature_selection(gamm.S2.seurat_norm)
human_F49B7.seurat_feat_select <- feature_selection(human_F49B7.seurat_norm)
human_IMR90.seurat_feat_select <- feature_selection(human_IMR90.seurat_norm)
saveRDS(gamm.S2.seurat_feat_select, file = paste0(output,"GAMM_S2_feat_select.rds"))
saveRDS(human_F49B7.seurat_feat_select, file = paste0(output,"human_F49B7_feat_select.rds"))
saveRDS(human_IMR90.seurat_feat_select, file = paste0(output,"human_IMR90_feat_select.rds"))
```

# Scaling data
```{r}
#gamm.S2.seurat_feat_select <- readRDS(file = "output_preprocess20230912_144047_cc/GAMM_S2_feat_select.rds")
gamm.S2.seurat_scaled <- scale_data(gamm.S2.seurat_feat_select)
human_F49B7.seurat_scaled <- scale_data(human_F49B7.seurat_feat_select)
human_IMR90.seurat_scaled <- scale_data(human_IMR90.seurat_feat_select)
```

# Dimensional reduction
```{r}
gamm.S2.seurat_dim_reduced <- run_and_visualize_pca(gamm.S2.seurat_scaled)
human_F49B7.seurat_dim_reduced <- run_and_visualize_pca(human_F49B7.seurat_scaled)
human_IMR90.seurat_dim_reduced <- run_and_visualize_pca(human_IMR90.seurat_scaled)
```
# Run umap
```{r}
gamm.S2.seurat_umap <- run_umap(gamm.S2.seurat_dim_reduced)
#human_D205.seurat_umap <- readRDS(file = "output_preprocess20230912_144047_cc/human_D205_umap.rds")
human_F49B7.seurat_umap <- run_umap(human_F49B7.seurat_dim_reduced)
human_IMR90.seurat_umap <- run_umap(human_IMR90.seurat_dim_reduced)
#saveRDS(human_D205.seurat_umap, file = paste0(output, "human_D205_umap.rds"))
```
# clustering
```{r}
gamm.S2.seurat_clustered <- perform_clustering(gamm.S2.seurat_umap)
human_F49B7.seurat_clustered <- perform_clustering(human_F49B7.seurat_umap)
human_IMR90.seurat_clustered <- perform_clustering(human_IMR90.seurat_umap)
saveRDS(gamm.S2.seurat_clustered, file = paste0(output, "gamm.S2_clustered.rds"))
saveRDS(human_F49B7.seurat_clustered, file = paste0(output, "human_F49B7_clustered.rds"))
saveRDS(human_IMR90.seurat_clustered, file = paste0(output, "human_IMR90_clustered.rds"))
```
# check metadata
```{r}
metadata1[1:5,]
metadata2[1:5,]
human_F49B7.seurat_clustered<-AddMetaData(object=human_F49B7.seurat_clustered, metadata=metadata1)
human_IMR90.seurat_clustered<-AddMetaData(object=human_IMR90.seurat_clustered, metadata=metadata2)
colnames(human_IMR90.seurat_clustered@meta.data)
table(human_IMR90.seurat_clustered$cell_type2)
colnames(human_F49B7.seurat_clustered@meta.data)
table(human_F49B7.seurat_clustered$cell_type2)
```
# save
```{r save}
# save objects
saveRDS(human_F49B7.seurat_clustered, file = paste0(output, "human_F49B7_clustered.rds"))
saveRDS(human_IMR90.seurat_clustered, file = paste0(output, "human_IMR90_clustered.rds"))
# save session info
writeLines(capture.output(sessionInfo()), paste0(output,"sessionInfo.txt"))
```