---
title: "preprocess_for_cross-species"
author: "Beth Moore"
date: "2023-07-07"
output: html_document
---
# Preprocessing for cross species analysis
# to have cross species, we need ortholog genes
# we then need to subset both reference and query by these genes
# reference data will also be subseted by its metadata

# Load libraries, set wd, and source functions
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(dplyr)
library(Seurat)
library(patchwork)
library(scran)
library(BiocParallel)
library(DropletUtils)
library(cowplot)
library(harmony)
library(SoupX)
library(scDblFinder)
library(reticulate)
library(purrr)
library(jsonlite)
library(rmarkdown)
BETHS_GAMM_DATA_DIR <- "/w5home/bmoore/scRNAseq/GAMM/"
use_condaenv("scRNAseq_best")
setwd(paste0(BETHS_GAMM_DATA_DIR, "human_data/cowan_cell_2020/"))
timestamp <- format(Sys.time(), "%Y%m%d_%H%M%S")
output <- paste0("output_preprocess_lanes_", timestamp)
dir.create(output, showWarnings = FALSE)
file.copy("/w5home/bmoore/scRNAseq_library/config.json", file.path(output, "config.json"))
config <- fromJSON(file.path(output, "config.json"))
output <- paste0(output, "/")
source("/w5home/bmoore/scRNAseq_library/sc_pipeline/src/sc_pipeline_functions.R")

```
# Read in data
```{r data}
# refernce data
exp.matrix1 <- read_aligned_data(
    paste0(BETHS_GAMM_DATA_DIR,"human_data/cowan_cell_2020/output_S408/Solo.out/GeneFull/"),
    "S408",output
    )
exp.matrix2 <- read_aligned_data(
    paste0(BETHS_GAMM_DATA_DIR,"human_data/cowan_cell_2020/output_S410/Solo.out/GeneFull/"),
    "S410",output
    )
exp.matrix3 <- read_aligned_data(
    paste0(BETHS_GAMM_DATA_DIR,"human_data/cowan_cell_2020/output_S411/Solo.out/GeneFull/"),
    "S411",output
    )
exp.matrix4 <- read_aligned_data(
    paste0(BETHS_GAMM_DATA_DIR,"human_data/cowan_cell_2020/output_S412/Solo.out/GeneFull/"),
    "S412",output
    )
exp.matrix5 <- read_aligned_data(
    paste0(BETHS_GAMM_DATA_DIR,"human_data/cowan_cell_2020/output_S413/Solo.out/GeneFull/"),
    "S413",output
    )
exp.matrix6 <- read_aligned_data(
    paste0(BETHS_GAMM_DATA_DIR,"human_data/cowan_cell_2020/output_S414/Solo.out/GeneFull/"),
    "S414",output
    )
```

# run soupx
```{r soupx}
all_aligned_data <- list(exp.matrix1,exp.matrix2,exp.matrix3,exp.matrix4,exp.matrix5,exp.matrix6)
soupX_objs <- purrr::map(all_aligned_data, ~prep_seurat_and_soupX(data.raw = .x$raw, data = .x$filtered, project = .x$project), .names = purrr::map_chr(all_data, 'project'))
```
# Create Seurat objects and convert to sce objects for doublet removal
```{r}
project_names <- purrr::map_chr(all_aligned_data, 'project')
feature_set1 <- list(feature1 = 'nCount_RNA', feature2 = 'nFeature_RNA')
filtered_list_of_objs <- purrr::map2(soupX_objs, project_names, ~create_seurat_and_sce(out = .x$out, project = .y, feature_set = feature_set1))

```
# Doublet removal and lane merging
```{r}
combine_feature_plots(filtered_list_of_objs, feature_set1, file_name_base = "post_soupx_qc")
filtered_sce_list <- list(filtered_list_of_objs[[1]]$sce, filtered_list_of_objs[[2]]$sce, filtered_list_of_objs[[3]]$sce, filtered_list_of_objs[[4]]$sce, filtered_list_of_objs[[5]]$sce, filtered_list_of_objs[[6]]$sce)
lane_and_merged_seurat_obj <- run_scDblFinder_and_merge(filtered_sce_list)
rm(filtered_list_of_objs, feature_set1, filtered_sce_list)
gc(full = TRUE)
```
# Mitochondrial gene filtering
```{r MT Gene Filtering}
seurat_obj_mt_filtered <- filter_cells(lane_and_merged_seurat_obj)
rm(lane_and_merged_seurat_obj)
gc(full = TRUE)
```

# Normalization
```{r Normalization}
normalized_seurat_obj <- normalize_data(seurat_obj_mt_filtered)
rm(seurat_obj_mt_filtered)
gc(full = TRUE)
```

# Feature selection
```{r Feature Selection}
feature_selected_seurat_obj <- feature_selection(normalized_seurat_obj)
rm(normalized_seurat_obj)
gc(full = TRUE)
```

# Scaling data
```{r Scaling Data}
scaled_seurat_obj <- scale_data(feature_selected_seurat_obj)
rm(feature_selected_seurat_obj)
gc(full = TRUE)
```

# Dimensional reduction
```{r Dimensional Reduction}
dim_reduced_seurat_obj <- run_and_visualize_pca(scaled_seurat_obj)
rm(scaled_seurat_obj)
gc(full = TRUE)
```

# Batch correction
```{r Batch Correction}
# Check the number of unique levels in orig.ident
if (length(unique(dim_reduced_seurat_obj$orig.ident)) > 1) {
  # Perform batch correction if orig.ident has more than one level
  batch_corrected_obj <- perform_batch_correction(dim_reduced_seurat_obj)
} else {
  # Skip batch correction and use the original object
  batch_corrected_obj <- dim_reduced_seurat_obj
  message("Skipping batch correction as orig.ident has only one level.")
}

# Remove the original object to free up memory
rm(dim_reduced_seurat_obj)
gc(full = TRUE)
```

# UMAP
```{r UMAP}
# Check if batch_corrected_obj is a list (i.e., batch correction was performed)
if (is.list(batch_corrected_obj)) {
  # Run UMAP on the Seurat object returned by perform_batch_correction
  umap_seurat_obj <- run_umap(batch_corrected_obj$seurat_obj)
} else {
  # Run UMAP on the original Seurat object (i.e., batch correction was skipped)
  umap_seurat_obj <- run_umap(batch_corrected_obj)
}

# Remove the batch_corrected_obj to free up memory
rm(batch_corrected_obj)
gc(full = TRUE)
```

# Clustering 
```{r Clustering}
clustered_seurat_obj <- perform_clustering(umap_seurat_obj)
rm(umap_seurat_obj)
gc(full = TRUE)
```

# add in metadata and visualize
```{r metadata}
# read in meta.data
# remember the set working directory in chunk 1, to imform this path
metadata <- read.csv2("meta_30weeks_mod.txt", sep="\t", header=TRUE, row.names=1)
# subset by D205 data
metadata1 <- subset(metadata, metadata$source=="f49b7")
metadata1<- as.data.frame(metadata1)
complete.cases(metadata1)
metadata1<-na.omit(metadata1)
table(metadata1$cell_type2)
# add to object
clustered_seurat_obj<-AddMetaData(object=clustered_seurat_obj, metadata=metadata1)
```

# Find markers with Differential expressed features (genes)
```{r Differential Expression}
if (config$DE_method == "Seurat") {
  de_results <- find_differentially_expressed_features(clustered_seurat_obj)
  analyze_known_markers(clustered_seurat_obj, de_results, output)
} else if (config$DE_method == "Scran") {
  annot_df <- score_and_plot_markers(clustered_seurat_obj, output)
}
```
# save the clustered object
```{r Save Final Object}
saveRDS(clustered_seurat_obj, file = paste0(output, "clustered_seurat_obj.rds"))
```
# manual annotation
```{r manual_annot}
# order clusters from annotation data frame
new_df.ordered <- annot_df[order(as.numeric(annot_df$Cluster)), ]
# get new cluster names ordered by cluster number
new.cluster.ids <- new_df.ordered$Cell.type
# annotate
labeled_seurat_obj <- annotate_clusters_and_save(clustered_seurat_obj, new.cluster.ids, output)
```