---
title: "scRNAseq_pipeline"
output: pdf_document
---

# ENVIRONMENT SETUP

```{r env, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(dplyr)
library(Seurat)
library(patchwork)
library(scran)
library(BiocParallel)
library(DropletUtils)
library(cowplot)
library(harmony)
library(SoupX)
library(scDblFinder)
library(reticulate)
library(purrr)
library(jsonlite)
library(rmarkdown)
use_condaenv("/root/miniconda/envs/scrnaseq")
timestamp <- format(Sys.time(), "%Y%m%d_%H%M%S")
output <- paste0("../output_", timestamp)
dir.create(output, showWarnings = FALSE)
file.copy("../config.json", file.path(output, "config.json"))
config <- jsonlite::fromJSON(file.path(output, "config.json"), simplifyDataFrame = FALSE)
output <- paste0(output, "/")
source("sc_pipeline_functions.R")
```

# Load Data
```{r load_data}
all_aligned_data <- lapply(config$lanes, function(lane) {
  read_aligned_data(lane$base_directory, lane$name)
})
```

# Data Prep
```{r SoupX}
soupX_objs <- purrr::map(all_aligned_data, ~ prep_seurat_and_soupX(data.raw = .x$raw, data = .x$filtered, project = .x$project), .names = purrr::map_chr(all_data, "project"))
```

# Create Seurat objects and convert to sce objects
```{r Object_Conversion}
project_names <- purrr::map_chr(all_aligned_data, "project")
feature_set1 <- list(feature1 = "nCount_RNA", feature2 = "nFeature_RNA")
filtered_list_of_objs <- purrr::map2(soupX_objs, project_names, ~ create_seurat_and_sce(out = .x$out, project = .y, feature_set = feature_set1))
```

# Doublet removal and lane merging
```{r Doublet Removal and Lane Merging}
combine_feature_plots(filtered_list_of_objs, feature_set1, file_name_base = "post_soupx_qc")
filtered_sce_list <- lapply(filtered_list_of_objs, function(x) x$sce)
lane_and_merged_seurat_obj <- run_scDblFinder_and_merge(filtered_sce_list)
```

# Mitochondrial gene filtering
```{r MT Gene Filtering}
seurat_obj_mt_filtered <- filter_cells(lane_and_merged_seurat_obj$seurat_obj)
```

# Normalization
```{r Normalization}
normalized_seurat_obj <- normalize_data(seurat_obj_mt_filtered)
```

# Feature selection
```{r Feature Selection}
feature_selected_seurat_obj <- feature_selection(normalized_seurat_obj)
```

# Scaling data
```{r Scaling Data}
scaled_seurat_obj <- scale_data(feature_selected_seurat_obj)
```

# Dimensional reduction
```{r Dimensional Reduction}
dim_reduced_seurat_obj <- run_and_visualize_pca(scaled_seurat_obj)
```

# Batch correction
```{r Batch Correction}
batch_corrected_obj <- perform_batch_correction(dim_reduced_seurat_obj)
```

# Clustering 
```{r Clustering}
clustered_seurat_obj <- perform_clustering(batch_corrected_obj$seurat_obj)
```

# Find markers with Differential expressed features (genes)
```{r Differential Expression}
if (config$DE_method == "Seurat") {
  de_results <- find_differentially_expressed_features(clustered_seurat_obj)
} else if (config$DE_method == "Scran") {
  de_results <- find_differentially_expressed_features_m3drop(clustered_seurat_obj)
}
```